File: library/src/main/java/me/shouheng/compress/strategy/compress/Compressor.java
package me.shouheng.compress.strategy.compress;

import android.graphics.*;
import android.os.AsyncTask;
import io.reactivex.Flowable;
import me.shouheng.compress.AbstractStrategy;
import me.shouheng.compress.strategy.config.Config;
import me.shouheng.compress.strategy.config.ScaleMode;
import me.shouheng.compress.utils.ImageUtils;
import me.shouheng.compress.utils.LogLog;

import java.io.ByteArrayOutputStream;
import java.io.File;
import java.io.FileOutputStream;
import java.io.IOException;
import java.util.concurrent.Callable;

/**
 * The compress algorithm by <a href="https://github.com/zetbaitsu/Compressor">Compressor</a>.
 *
 * @author WngShhng
 * @version 2019-05-17
 */
public class Compressor extends AbstractStrategy {

    private float maxWidth = Config.COMPRESSOR_DEFAULT_MAX_WIDTH;

    private float maxHeight = Config.COMPRESSOR_DEFAULT_MAX_HEIGHT;

    @ScaleMode.Mode
    private int scaleMode = Config.COMPRESSOR_DEFAULT_SCALE_MODE;

    private Bitmap.Config config;

    /*--------------------------------------- public methods ------------------------------------------*/

    /**
     * Set the max width of compressed image.
     *
     * @param maxWidth the max width in pixels.
     * @return compressor object.
     */
    public Compressor setMaxWidth(float maxWidth) {
        this.maxWidth = maxWidth;
        return this;
    }

    /**
     * Set the max height of compressed image.
     *
     * @param maxHeight the max height in pixels.
     * @return the compressor object.
     */
    public Compressor setMaxHeight(float maxHeight) {
        this.maxHeight = maxHeight;
        return this;
    }

    /**
     * Set the scale mode when the destination image ratio differ from the original original.
     * Might be one of {@link ScaleMode#SCALE_LARGER},
     * {@link ScaleMode#SCALE_SMALLER},
     * {@link ScaleMode#SCALE_WIDTH} or
     * {@link ScaleMode#SCALE_HEIGHT}.
     *
     * @param scaleMode the scale mode.
     * @return the compressor object.
     * @see ScaleMode
     */
    public Compressor setScaleMode(@ScaleMode.Mode int scaleMode) {
        this.scaleMode = scaleMode;
        return this;
    }

    /**
     * Set the image configuration for bitmap: {@link android.graphics.Bitmap.Config}.
     *
     * @param config the config
     * @return the compress object
     * @see android.graphics.Bitmap.Config
     */
    public Compressor setConfig(Bitmap.Config config) {
        this.config = config;
        return this;
    }

    @Override
    public File get() {
        try {
            notifyCompressStart();
            compressAndWrite();
            notifyCompressSuccess(outFile);
        } catch (Exception e) {
            LogLog.e(e.getMessage());
            notifyCompressError(e);
        }
        return outFile;
    }

    @Override
    public Flowable<File> asFlowable() {
        return Flowable.defer(new Callable<Flowable<File>>() {
            @Override
            public Flowable<File> call() {
                try {
                    notifyCompressStart();
                    boolean succeed = compressAndWrite();
                    if (succeed) {
                        notifyCompressSuccess(outFile);
                    } else {
                        notifyCompressError(new Exception("Failed to compress image, either caused by OOM or other problems."));
                    }
                    return Flowable.just(outFile);
                } catch (Exception e) {
                    notifyCompressError(e);
                    LogLog.e(e.getMessage());
                    return Flowable.error(e);
                }
            }
        });
    }

    @Override
    public void launch() {
        AsyncTask.SERIAL_EXECUTOR.execute(new Runnable() {
            @Override
            public void run() {
                try {
                    notifyCompressStart();
                    boolean succeed = compressAndWrite();
                    if (succeed) {
                        notifyCompressSuccess(outFile);
                    } else {
                        notifyCompressError(new Exception("Failed to compress image, either caused by OOM or other problems."));
                    }
                } catch (Exception e) {
                    notifyCompressError(e);
                    LogLog.e(e.getMessage());
                }
            }
        });
    }

    @Override
    protected Bitmap getBitmap() {
        return compressByQuality();
    }

    /*--------------------------------------- protected methods ------------------------------------------*/

    protected int calculateRequiredWidth(float imgRatio, float reqRatio) {
        switch (scaleMode) {
            case ScaleMode.SCALE_LARGER:
                if (srcHeight > maxHeight || srcWidth > maxWidth) {
                    // If Height is greater
                    if (imgRatio < reqRatio) {
                        imgRatio = maxHeight / srcHeight;
                        return (int) (imgRatio * srcWidth);
                    }  // If Width is greater
                    else if (imgRatio > reqRatio) {
                        return (int) maxWidth;
                    }
                }
                break;
            case ScaleMode.SCALE_SMALLER:
                if (srcHeight > maxHeight || srcWidth > maxWidth) {
                    // If Height is greater
                    if (imgRatio < reqRatio) {
                        return (int) maxWidth;
                    }  // If Width is greater
                    else if (imgRatio > reqRatio) {
                        imgRatio = maxHeight / srcHeight;
                        return (int) (imgRatio * srcWidth);
                    }
                }
                break;
            case ScaleMode.SCALE_HEIGHT:
                return (int) (srcWidth * maxHeight / srcHeight);
            case ScaleMode.SCALE_WIDTH:
                return (int) maxWidth;
            default:
                return (int) maxWidth;
        }
        return (int) maxWidth;
    }

    protected int calculateRequiredHeight(float imgRatio, float reqRatio) {
        switch (scaleMode) {
            case ScaleMode.SCALE_LARGER:
                if (srcHeight > maxHeight || srcWidth > maxWidth) {
                    // If Height is greater
                    if (imgRatio < reqRatio) {
                        return (int) maxHeight;
                    }  // If Width is greater
                    else if (imgRatio > reqRatio) {
                        imgRatio = maxWidth / srcWidth;
                        return (int) (imgRatio * srcHeight);
                    }
                }
                break;
            case ScaleMode.SCALE_SMALLER:
                if (srcHeight > maxHeight || srcWidth > maxWidth) {
                    // If Height is greater
                    if (imgRatio < reqRatio) {
                        imgRatio = maxWidth / srcWidth;
                        return (int) (imgRatio * srcHeight);
                    }  // If Width is greater
                    else if (imgRatio > reqRatio) {
                        return (int) maxHeight;
                    }
                }
                break;
            case ScaleMode.SCALE_HEIGHT:
                return (int) maxHeight;
            case ScaleMode.SCALE_WIDTH:
                imgRatio = maxWidth / srcWidth;
                return (int) (srcHeight * imgRatio);
            default:
                return (int) maxHeight;
        }
        return (int) maxHeight;
    }

    protected int calculateInSampleSize(BitmapFactory.Options options, int reqWidth, int reqHeight) {
        // Raw height and width of image
        final int height = options.outHeight;
        final int width = options.outWidth;
        int inSampleSize = 1;

        if (height > reqHeight || width > reqWidth) {
            final int halfHeight = height / 2;
            final int halfWidth = width / 2;

            // Calculate the largest inSampleSize value that is a power of 2 and keeps both
            // height and width larger than the requested height and width.
            while ((halfHeight / inSampleSize) >= reqHeight && (halfWidth / inSampleSize) >= reqWidth) {
                inSampleSize *= 2;
            }
        }

        return inSampleSize;
    }

    /*--------------------------------------- inner methods ------------------------------------------*/

    /**
     * Compress bitmap and save it to target file.
     *
     * @return is the full compress logic succeed.
     * @throws IOException io exception
     */
    private boolean compressAndWrite() throws IOException {
        Bitmap bitmap = compressByScale();
        if (!ImageUtils.isEmptyBitmap(bitmap)) {
            FileOutputStream fos = new FileOutputStream(outFile);
            bitmap.compress(format, quality, fos);
            fos.flush();
            fos.close();
        } else {
            return false;
        }
        return true;
    }

    /**
     * Compress by quality, the bitmap will be compressed by scale first.
     *
     * @return the compressed bitmap
     */
    private Bitmap compressByQuality() {
        Bitmap bitmap = compressByScale();
        if (ImageUtils.isEmptyBitmap(bitmap)) return null;
        ByteArrayOutputStream baos = new ByteArrayOutputStream();
        bitmap.compress(format, quality, baos);
        byte[] bytes = baos.toByteArray();
        return BitmapFactory.decodeByteArray(bytes, 0, bytes.length);
    }

    /**
     * Compress the source bitmap by scale.
     *
     * @return the compressed bitmap
     */
    private Bitmap compressByScale() {
        prepareImageSizeInfo();

        float imgRatio = (float) srcWidth / (float) srcHeight;
        float reqRatio = maxWidth / maxHeight;

        int reqWidth = calculateRequiredWidth(imgRatio, reqRatio);
        int reqHeight = calculateRequiredHeight(imgRatio, reqRatio);
        Bitmap bmp = srcBitmap;

        if (srcData != null || srcFile != null) {
            BitmapFactory.Options options = new BitmapFactory.Options();
            options.inJustDecodeBounds = true;
            options.inSampleSize = 1;

            if (srcData != null) {
                BitmapFactory.decodeByteArray(srcData, 0, srcData.length, options);
            } else {
                BitmapFactory.decodeFile(srcFile.getAbsolutePath(), options);
            }

            options.inSampleSize = calculateInSampleSize(options, reqWidth, reqHeight);
            options.inJustDecodeBounds = false;
            options.inDither = false;
            options.inPurgeable = true;
            options.inInputShareable = true;
            options.inTempStorage = new byte[16 * 1024];

            if (srcData != null) {
                bmp = BitmapFactory.decodeByteArray(srcData, 0, srcData.length, options);
            } else {
                bmp = BitmapFactory.decodeFile(srcFile.getAbsolutePath(), options);
            }
        }

        Bitmap scaledBitmap = Bitmap.createBitmap(reqWidth, reqHeight, config == null ? Bitmap.Config.ARGB_8888 : config);

        // return null if OOM.
        if (bmp == null || scaledBitmap == null) {
            return null;
        }

        // scale the bitmap.
        float ratioX = reqWidth / (float) bmp.getWidth();
        float ratioY = reqHeight / (float) bmp.getHeight();
        float middleX = reqWidth / 2.0f;
        float middleY = reqHeight / 2.0f;

        Matrix scaleMatrix = new Matrix();
        scaleMatrix.setScale(ratioX, ratioY, middleX, middleY);

        Canvas canvas = new Canvas(scaledBitmap);
        canvas.setMatrix(scaleMatrix);
        canvas.drawBitmap(bmp, middleX - bmp.getWidth() / 2,
                middleY - bmp.getHeight() / 2, new Paint(Paint.FILTER_BITMAP_FLAG));
        bmp.recycle();

        if (srcFile != null) {
            int orientation = ImageUtils.getImageAngle(srcFile);
            if (orientation != 0) {
                scaledBitmap = ImageUtils.rotateBitmap(scaledBitmap, ImageUtils.getImageAngle(srcFile));
            }
        }

        return scaledBitmap;
    }

}

File: library/src/main/java/me/shouheng/compress/strategy/compress/Compressor.kt
package me.shouheng.compress.strategy.compress

import android.graphics.Bitmap
import android.graphics.BitmapFactory
import android.graphics.Matrix
import android.os.AsyncTask
import io.reactivex.Flowable
import kotlinx.coroutines.withContext
import me.shouheng.compress.AbstractStrategy
import me.shouheng.compress.strategy.config.Config
import me.shouheng.compress.strategy.config.ScaleMode
import me.shouheng.compress.utils.CLog
import me.shouheng.compress.utils.isEmpty
import me.shouheng.compress.utils.isNotEmpty
import me.shouheng.compress.utils.rotate
import java.io.ByteArrayOutputStream
import java.io.File
import java.io.FileOutputStream
import java.util.concurrent.Callable
import kotlin.coroutines.CoroutineContext

/** The concrete compress algorithm. */
open class Compressor : AbstractStrategy() {

    private var maxWidth: Float             = Config.COMPRESSOR_DEFAULT_MAX_WIDTH
    private var maxHeight: Float            = Config.COMPRESSOR_DEFAULT_MAX_HEIGHT
    @ScaleMode private var scaleMode: Int   = Config.COMPRESSOR_DEFAULT_SCALE_MODE
    private var config: Bitmap.Config?      = null
    private var ignoreIfSmaller: Boolean    = true

    override fun getBitmap(): Bitmap? = compressByScaleAndQuality()

    /*--------------------------------------- public methods ------------------------------------------*/

    /** Set the max width of compressed image in pixels. */
    fun setMaxWidth(maxWidth: Float): Compressor {
        this.maxWidth = maxWidth
        return this
    }

    /** Set the max height of compressed image in pixels. */
    fun setMaxHeight(maxHeight: Float): Compressor {
        this.maxHeight = maxHeight
        return this
    }

    /** Set the scale mode when the destination image ratio differ from the original original. */
    fun setScaleMode(@ScaleMode scaleMode: Int): Compressor {
        this.scaleMode = scaleMode
        return this
    }

    /** Set the image configuration for bitmap: [android.graphics.Bitmap.Config]. */
    fun setConfig(config: Bitmap.Config): Compressor {
        this.config = config
        return this
    }

    /** Don't compress if the desired bitmap is smaller than desired size. */
    fun setIgnoreIfSmaller(ignoreIfSmaller: Boolean): Compressor {
        this.ignoreIfSmaller = ignoreIfSmaller
        return this
    }

    override fun get(): File {
        try {
            notifyCompressStart()
            compressAndWrite()
            notifyCompressSuccess(outFile!!)
        } catch (e: Exception) {
            CLog.e(e.message)
            notifyCompressError(e)
        }
        return outFile!!
    }

    override suspend fun get(coroutineContext: CoroutineContext): File =
        withContext(coroutineContext) {
            return@withContext get()
        }

    override fun asFlowable(): Flowable<File> {
        return Flowable.defer(Callable {
            try {
                notifyCompressStart()
                val succeed = compressAndWrite()
                if (succeed) {
                    notifyCompressSuccess(outFile!!)
                } else {
                    notifyCompressError(Exception("Failed to compress image, " +
                            "either caused by OOM or other problems."))
                }
                return@Callable Flowable.just(outFile)
            } catch (e: Exception) {
                notifyCompressError(e)
                CLog.e(e.message)
                return@Callable Flowable.error<File>(e)
            }
        })
    }

    override fun launch() {
        AsyncTask.SERIAL_EXECUTOR.execute {
            try {
                notifyCompressStart()
                val succeed = compressAndWrite()
                if (succeed) {
                    notifyCompressSuccess(outFile!!)
                } else {
                    notifyCompressError(Exception("Failed to compress image, " +
                            "either caused by OOM or other problems."))
                }
            } catch (e: Exception) {
                notifyCompressError(e)
                CLog.e(e.message)
            }
        }
    }

    /*--------------------------------------- protected methods ------------------------------------------*/

    open fun calculateRequiredWidth(imgRatio: Float, reqRatio: Float): Int {
        var ratio = imgRatio
        when (scaleMode) {
            ScaleMode.SCALE_LARGER -> if (srcHeight > maxHeight || srcWidth > maxWidth) {
                // If Height is greater
                if (ratio < reqRatio) {
                    ratio = maxHeight / srcHeight
                    return (ratio * srcWidth).toInt()
                }  // If Width is greater
                else if (ratio > reqRatio) {
                    return maxWidth.toInt()
                }
            }
            ScaleMode.SCALE_SMALLER -> if (srcHeight > maxHeight || srcWidth > maxWidth) {
                // If Height is greater
                if (ratio < reqRatio) {
                    return maxWidth.toInt()
                }  // If Width is greater
                else if (ratio > reqRatio) {
                    ratio = maxHeight / srcHeight
                    return (ratio * srcWidth).toInt()
                }
            }
            ScaleMode.SCALE_HEIGHT -> return (srcWidth * maxHeight / srcHeight).toInt()
            ScaleMode.SCALE_WIDTH -> return maxWidth.toInt()
            else -> return maxWidth.toInt()
        }
        return maxWidth.toInt()
    }

    open fun calculateRequiredHeight(imgRatio: Float, reqRatio: Float): Int {
        var ratio = imgRatio
        when (scaleMode) {
            ScaleMode.SCALE_LARGER -> if (srcHeight > maxHeight || srcWidth > maxWidth) {
                // If Height is greater
                if (ratio < reqRatio) {
                    return maxHeight.toInt()
                }  // If Width is greater
                else if (ratio > reqRatio) {
                    ratio = maxWidth / srcWidth
                    return (ratio * srcHeight).toInt()
                }
            }
            ScaleMode.SCALE_SMALLER -> if (srcHeight > maxHeight || srcWidth > maxWidth) {
                // If Height is greater
                if (ratio < reqRatio) {
                    ratio = maxWidth / srcWidth
                    return (ratio * srcHeight).toInt()
                }  // If Width is greater
                else if (ratio > reqRatio) {
                    return maxHeight.toInt()
                }
            }
            ScaleMode.SCALE_HEIGHT -> return maxHeight.toInt()
            ScaleMode.SCALE_WIDTH -> {
                ratio = maxWidth / srcWidth
                return (srcHeight * ratio).toInt()
            }
            else -> return maxHeight.toInt()
        }
        return maxHeight.toInt()
    }

    open fun calculateInSampleSize(reqWidth: Int, reqHeight: Int): Int {
        var inSampleSize = 1
        if (srcHeight > reqHeight || srcWidth > reqWidth) {
            val halfHeight = srcHeight / 2
            val halfWidth = srcWidth / 2
            // Calculate the largest inSampleSize value that is a power of 2 and keeps both
            // height and width larger than the requested height and width.
            while (halfHeight / inSampleSize >= reqHeight && halfWidth / inSampleSize >= reqWidth) {
                inSampleSize *= 2
            }
        }
        return inSampleSize
    }

    /*--------------------------------------- inner methods ------------------------------------------*/

    /** Compress bitmap and save it to target file. */
    private fun compressAndWrite(): Boolean {
        val bitmap = compressByScale()
        if (bitmap.isNotEmpty()) {
            val fos = FileOutputStream(outFile)
            bitmap?.compress(format, quality, fos)
            fos.flush()
            fos.close()
        } else {
            return false
        }
        return true
    }

    /** Compress by quality. The bitmap will be compressed by scale first. */
    private fun compressByScaleAndQuality(): Bitmap? {
        val bitmap = compressByScale()
        if (bitmap.isEmpty()) return null
        val baos = ByteArrayOutputStream()
        bitmap?.compress(format, quality, baos)
        val bytes = baos.toByteArray()
        return BitmapFactory.decodeByteArray(bytes, 0, bytes.size)
    }

    /** Compress the source bitmap by scale */
    private fun compressByScale(): Bitmap? {
        prepareImageSizeInfo()

        val imgRatio = srcWidth.toFloat() / srcHeight.toFloat()
        val reqRatio = maxWidth / maxHeight

        val reqWidth = calculateRequiredWidth(imgRatio, reqRatio)
        val reqHeight = calculateRequiredHeight(imgRatio, reqRatio)

        val origin = getOriginBitmap(reqWidth, reqHeight)

        // ignore if the origin size is smaller then required size
        if (ignoreIfSmaller && (reqWidth > srcWidth || reqWidth > srcHeight)) {
            return getRotatedBitmap(origin)
        }

        // failed to create scaled bitmap or get the origin bitmap, return null directly.
        if (origin == null) return null

        // scale the bitmap
        val ratioX = reqWidth / origin.width.toFloat()
        val ratioY = reqHeight / origin.height.toFloat()
        val matrix = Matrix().apply { setScale(ratioX, ratioY) }
        val result = Bitmap.createBitmap(origin, 0, 0, origin.width, origin.height, matrix, true)

        return getRotatedBitmap(result)
    }

    /** Get rotated image bitmap if necessary. */
    private fun getRotatedBitmap(bitmap: Bitmap?): Bitmap? {
        if (bitmap == null) return bitmap
        val rotation = imageSource?.getRotation()
        return if (rotation != null && rotation != 0) bitmap.rotate(rotation) else bitmap
    }

    /** Get the origin bitmap based on the source bitmap type. */
    private fun getOriginBitmap(reqWidth: Int, reqHeight: Int): Bitmap? {
        val options = BitmapFactory.Options()
        options.inSampleSize = calculateInSampleSize(reqWidth, reqHeight)
        options.inJustDecodeBounds = false
        options.inDither = false
        options.inPurgeable = true
        options.inInputShareable = true
        options.inTempStorage = ByteArray(16 * 1024)
        return imageSource?.getOriginBitmapByOptions(options)
    }
}
